README.md
# Angle

Simple JavaScript library for working with angles.

## Documentation

### Angle

Class representing an angle.

#### constructor

Create an angle.

**Parameters**

-   `config` **[Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)**
    -   `config.directions_start` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** Value with which first direction keyword will be associated.
    -   `config.directions` **[Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)>** List of direction keywords.
-   `$0` **[Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)**  (optional, default `{}`)
    -   `$0.directions_start`   (optional, default `default_directions_start`)
    -   `$0.directions`   (optional, default `default_directions`)

#### directions_start

#### directions

#### toDirection

Converts angle in degrees to direction keyword.

**Parameters**

-   `angle` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**

**Examples**

```javascript
(new Angle()).toDirection(90); // --> 'top'
```

Returns **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)**

#### fromDirection

Converts direction keyword to angle in degrees. Returns `null` if direction keyword is not defined.

**Parameters**

-   `direction` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)**

**Examples**

```javascript
(new Angle()).fromDirection('top'); // --> 90
```

Returns **([number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number) | null)**

#### radToDeg

Convert angle from radians to degrees.

**Parameters**

-   `angle` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**

**Examples**

```javascript
Array.radToDeg(Math.PI); // --> 180
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**

#### degToRad

Convert angle from degrees to radians.

**Parameters**

-   `angle` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**

**Examples**

```javascript
Array.radToDeg(180); // --> Math.PI
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**

#### normalize

Converts angle in degrees into equal angle in range between 0 and 360.

**Parameters**

-   `angle` **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**

**Examples**

```javascript
Array.normalize(450); // --> 90
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)**

## Bug reports, feature requests and contact

If you found any bugs, if you have feature requests or any questions, please, either [file an issue at GitHub](https://github.com/fczbkk/angle-js/issues) or send me an e-mail at <a href="mailto:riki@fczbkk.com">riki@fczbkk.com</a>.

## License

Angle is published under the [MIT license](https://github.com/fczbkk/angle-js/blob/master/LICENSE).


CHANGELOG.md
<a name="2.0.1"></a>
## [2.0.1](https://github.com/fczbkk/angle-js/compare/v2.0.0...v2.0.1) (2016-06-03)


### Bug Fixes

* use correct path to GIT ([6a133dd](https://github.com/fczbkk/angle-js/commit/6a133dd))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/fczbkk/angle/compare/v1.1.0...v2.0.0) (2016-06-01)





test/index.spec.js
import Angle from './../src/';


describe('Angle', function () {

  it('should exist', function () {
    expect(Angle).toBeDefined();
  });

  describe('config', function () {

    it('should be created with default directions', function () {
      const x = new Angle();
      expect(x.directions).toEqual([
        'right', 'topRight', 'top', 'topLeft',
        'left', 'bottomLeft', 'bottom', 'bottomRight'
      ]);
    });

    it('should be created with default directions_start', function () {
      const x = new Angle();
      expect(x.directions_start).toEqual(0);
    });

    it('should not allow to set invalid directions', function () {
      const x = new Angle();
      const fn = function () {x.directions = 'xxx';};
      expect(fn).toThrow();
    });

    it('should not allow to set invalid directions_start', function () {
      const x = new Angle();
      const fn = function () {x.directions_start = 'xxx';};
      expect(fn).toThrow();
    });

  });

  describe('DEG-RAD conversion', function () {

    it('should convert from DEG to RAD', function () {
      expect(Angle.radToDeg(0)).toEqual(0);
      expect(Angle.radToDeg(Math.PI)).toEqual(180);
    });

    it('should convert from RAD to DEG', function () {
      expect(Angle.degToRad(0)).toEqual(0);
      expect(Angle.degToRad(180)).toEqual(Math.PI);
    });

  });

  describe('normalization', function () {

    it('should keep value in range between 0 and 360 unchanged', function () {
      expect(Angle.normalize(0)).toEqual(0);
      expect(Angle.normalize(180)).toEqual(180);
    });

    it('should convert negative value', function () {
      expect(Angle.normalize(-90)).toEqual(270);
      expect(Angle.normalize(-180)).toEqual(180);
    });

    it('should convert large value', function () {
      expect(Angle.normalize(360)).toEqual(0);
      expect(Angle.normalize(450)).toEqual(90);
    });

    it('should convert to 0 on invalid value', function () {
      expect(Angle.normalize(null)).toEqual(0);
      expect(Angle.normalize('xxx')).toEqual(0);
    });

  });

  describe('angle to direction', function () {

    it('should get direction by angle', function () {
      const x = new Angle();
      expect(x.toDirection(0)).toEqual('right');
      expect(x.toDirection(90)).toEqual('top');
    });

    it('should use custom directions', function () {
      const x = new Angle({directions: ['right', 'left']});
      expect(x.toDirection(0)).toEqual('right');
      expect(x.toDirection(89)).toEqual('right');
      expect(x.toDirection(91)).toEqual('left');
      expect(x.toDirection(180)).toEqual('left');
    });

    it('should adjust by direction start', function () {
      const x = new Angle({
        directions: ['up', 'down'],
        directions_start: 90
      });
      expect(x.toDirection(90)).toEqual('up');
      expect(x.toDirection(45)).toEqual('up');
      expect(x.toDirection(135)).toEqual('up');
      expect(x.toDirection(270)).toEqual('down');
      expect(x.toDirection(225)).toEqual('down');
      expect(x.toDirection(315)).toEqual('down');
    });

  });

  describe('direction to angle', function () {

    it('should convert direction to angle', function () {
      const x = new Angle();
      expect(x.fromDirection('right')).toEqual(0);
      expect(x.fromDirection('top')).toEqual(90);
      expect(x.fromDirection('left')).toEqual(180);
    });

    it('should use custom directions', function () {
      const x = new Angle({directions: ['right', 'left']});
      expect(x.fromDirection('right')).toEqual(0);
      expect(x.fromDirection('left')).toEqual(180);
    });

    it('should use custom direction start', function () {
      const x = new Angle({
        directions: ['up', 'down'],
        directions_start: 90
      });
      expect(x.fromDirection('up')).toEqual(90);
      expect(x.fromDirection('down')).toEqual(270);
    });

    it('should return `null` if direction does not exist', function () {
      const x = new Angle();
      expect(x.fromDirection('xxx')).toEqual(null);
    });

  });

});


src/index.js
const default_directions_start = 0;
const default_directions = [
  'right',
  'topRight',
  'top',
  'topLeft',
  'left',
  'bottomLeft',
  'bottom',
  'bottomRight'
];


/**
 * Class representing an angle.
 * @name Angle
 */
export default class Angle {

  /**
   * Create an angle.
   * @param {Object} config
   * @param {number} config.directions_start Value with which first direction keyword will be associated.
   * @param {Array.<string>} config.directions List of direction keywords.
   */
  constructor ({
    directions_start = default_directions_start,
    directions = default_directions
  } = {}) {
    this.directions_start = directions_start;
    this.directions = directions;
  }

  /**
   * @type {number}
   */
  get directions_start () {
    return this._directions_start;
  }

  set directions_start (directions_start) {
    if (typeof directions_start === 'number') {
      this._directions_start = directions_start;
    } else {
      throw new TypeError('Angle: directions_start must be a number.');
    }
  }

  /**
   * @type {Array}
   */
  get directions () {
    return this._directions;
  }

  set directions (directions) {
    if (Array.isArray(directions)) {
      this._directions = directions;
    } else {
      throw new TypeError('Angle: directions must by an array.');
    }
  }

  /**
   * Convert angle from radians to degrees.
   * @param {number} angle
   * @returns {number}
   * @static
   * @example Array.radToDeg(Math.PI); // --> 180
   */
  static radToDeg (angle) {
    return angle * (180 / Math.PI);
  }

  /**
   * Convert angle from degrees to radians.
   * @param {number} angle
   * @returns {number}
   * @static
   * @example Array.radToDeg(180); // --> Math.PI
   */
  static degToRad (angle) {
    return angle * (Math.PI / 180);
  }

  /**
   * Converts angle in degrees into equal angle in range between 0 and 360.
   * @param {number} angle
   * @returns {number}
   * @static
   * @example Array.normalize(450); // --> 90
   */
  static normalize (angle) {
    let result = 0;
    if (typeof angle === 'number') {
      result = angle % 360;
      if (result < 0) {
        result += 360
      }
    }
    return result;
  }

  /**
   * Converts angle in degrees to direction keyword.
   * @param {number} angle
   * @returns {string}
   * @example (new Angle()).toDirection(90); // --> 'top'
   */
  toDirection (angle) {
    angle = angle - this.directions_start;
    angle = Angle.normalize(angle);
    const index = Math.round(angle / (360 / this.directions.length));
    return (typeof this.directions[index] === 'undefined')
      ? this.directions[0]
      : this.directions[index];
  }

  /**
   * Converts direction keyword to angle in degrees. Returns `null` if direction keyword is not defined.
   * @param {string} direction
   * @returns {number|null}
   * @example (new Angle()).fromDirection('top'); // --> 90
   */
  fromDirection (direction) {
    const index = this.directions.indexOf(direction);

    if (index !== -1) {
      let angle = index * (360 / this.directions.length);
      angle = this.directions_start + angle;
      return angle;
    }

    return null;
  }

}
